File: packages/shared/src/scoring.ts
import type { MetricsPoint, ScoreBreakdown, SessionInput, TranscriptSegment } from "./types";

const FILLER_WORDS = [
  "um",
  "uh",
  "like",
  "you know",
  "actually",
  "basically",
  "literally",
];

const STRUCTURE_WORDS = [
  "first",
  "second",
  "third",
  "because",
  "therefore",
  "however",
  "so",
  "thus",
];

const tokenize = (text: string) =>
  text
    .toLowerCase()
    .replace(/[^a-z0-9\s']/g, " ")
    .split(/\s+/)
    .filter(Boolean);

export const countWords = (segments: TranscriptSegment[]) =>
  segments.reduce((acc, seg) => acc + tokenize(seg.text).length, 0);

export const countFillers = (segments: TranscriptSegment[]) => {
  const text = segments.map((seg) => seg.text.toLowerCase()).join(" ");
  let count = 0;
  for (const filler of FILLER_WORDS) {
    const pattern = new RegExp(`\\b${filler.replace(" ", "\\s+")}\\b`, "g");
    const matches = text.match(pattern);
    if (matches) {
      count += matches.length;
    }
  }
  return count;
};

export const computeWpm = (wordCount: number, durationMs: number) => {
  if (durationMs <= 0) return 0;
  return Math.round((wordCount / (durationMs / 60000)) * 10) / 10;
};

const scoreWpm = (wpm: number) => {
  if (wpm >= 120 && wpm <= 160) return 18;
  if ((wpm >= 100 && wpm < 120) || (wpm > 160 && wpm <= 180)) return 10;
  return 4;
};

const scoreFillers = (fillersPerMinute: number) => {
  if (fillersPerMinute <= 2) return 12;
  if (fillersPerMinute <= 6) return 8;
  return 3;
};

const scorePauses = (pauseEvents: number[], durationMs: number) => {
  if (!pauseEvents.length) {
    return 4;
  }
  const natural = pauseEvents.filter((p) => p >= 600 && p <= 1500).length;
  const long = pauseEvents.filter((p) => p > 1500).length;
  const minutes = Math.max(durationMs / 60000, 0.1);
  const rate = natural / minutes;
  if (long > Math.max(2, minutes)) {
    return 4;
  }
  if (rate >= 2 && rate <= 6) return 10;
  if (rate >= 1) return 7;
  return 5;
};

const scoreEyeContact = (eyeContactPct: number) => {
  if (eyeContactPct > 0.7) return 15;
  if (eyeContactPct >= 0.4) return 10;
  return 5;
};

const scoreSmile = (smileAvg: number) => Math.round(7 * Math.min(1, Math.max(smileAvg, 0)));

const scoreHeadStability = (yawVar: number, pitchVar: number, rollVar: number) => {
  const totalVar = yawVar + pitchVar + rollVar;
  if (totalVar < 5) return 8;
  if (totalVar < 12) return 6;
  return 4;
};

const scoreStructure = (segments: TranscriptSegment[]) => {
  const words = tokenize(segments.map((s) => s.text).join(" "));
  const count = words.filter((w) => STRUCTURE_WORDS.includes(w)).length;
  if (count >= 4) return 10;
  if (count >= 2) return 7;
  return 4;
};

const scoreRepetition = (segments: TranscriptSegment[]) => {
  const words = tokenize(segments.map((s) => s.text).join(" "));
  const freq = new Map<string, number>();
  for (const word of words) {
    freq.set(word, (freq.get(word) ?? 0) + 1);
  }
  const repeated = Array.from(freq.values()).filter((count) => count >= 6).length;
  if (repeated >= 3) return -8;
  if (repeated >= 1) return -4;
  return 0;
};

const scoreClarity = (segments: TranscriptSegment[]) => {
  const sentences = segments
    .map((s) => s.text)
    .join(" ")
    .split(/[.!?]+/)
    .map((s) => s.trim())
    .filter(Boolean);
  if (!sentences.length) return 0;
  const lengths = sentences.map((s) => tokenize(s).length);
  const avg = lengths.reduce((a, b) => a + b, 0) / lengths.length;
  if (avg >= 8 && avg <= 20) return 12;
  if (avg >= 5 && avg <= 25) return 8;
  return 4;
};

const scoreKeywordCoverage = (segments: TranscriptSegment[], keywords: string[]) => {
  if (!keywords.length) return 0;
  const text = segments.map((s) => s.text.toLowerCase()).join(" ");
  const covered = keywords.filter((k) => text.includes(k.toLowerCase())).length;
  const ratio = covered / keywords.length;
  if (ratio >= 0.8) return 8;
  if (ratio >= 0.5) return 5;
  return 2;
};

const avg = (values: number[]) =>
  values.length ? values.reduce((a, b) => a + b, 0) / values.length : 0;

const variance = (values: number[]) => {
  if (!values.length) return 0;
  const mean = avg(values);
  return avg(values.map((v) => (v - mean) ** 2));
};

const collectPauseEvents = (metrics: MetricsPoint[]) =>
  metrics.map((m) => m.pauseMs).filter((p) => p > 0);

export const computeScore = (input: SessionInput): ScoreBreakdown => {
  const wordCount = countWords(input.transcript);
  const wpm = computeWpm(wordCount, input.durationMs);
  const fillerCount = countFillers(input.transcript);
  const minutes = Math.max(input.durationMs / 60000, 0.1);
  const fillersPerMinute = fillerCount / minutes;
  const pauseEvents = collectPauseEvents(input.metrics);

  const wpmScore = scoreWpm(wpm);
  const fillerScore = scoreFillers(fillersPerMinute);
  const pauseScore = scorePauses(pauseEvents, input.durationMs);
  const speech = wpmScore + fillerScore + pauseScore;

  const eyeContactPct = avg(input.metrics.map((m) => m.eyeContact));
  const smileAvg = avg(input.metrics.map((m) => m.smile));
  const yawVar = variance(input.metrics.map((m) => m.yaw));
  const pitchVar = variance(input.metrics.map((m) => m.pitch));
  const rollVar = variance(input.metrics.map((m) => m.roll));

  const eyeScore = scoreEyeContact(eyeContactPct);
  const smileScore = scoreSmile(smileAvg);
  const headScore = scoreHeadStability(yawVar, pitchVar, rollVar);
  const delivery = eyeScore + smileScore + headScore;

  const structureScore = scoreStructure(input.transcript);
  const repetitionScore = scoreRepetition(input.transcript);
  const clarityScore = scoreClarity(input.transcript);
  const keywordScore = scoreKeywordCoverage(input.transcript, input.targetKeywords);
  const content = Math.max(0, structureScore + repetitionScore + clarityScore + keywordScore);

  const total = Math.min(100, Math.max(0, speech + delivery + content));

  const fixes: string[] = [];
  if (wpmScore <= 10) fixes.push("Adjust pace to 120-160 WPM.");
  if (fillerScore <= 8) fixes.push("Reduce filler words with pauses.");
  if (eyeScore <= 10) fixes.push("Increase eye contact with the lens.");
  if (clarityScore <= 8) fixes.push("Shorten sentences for clarity.");
  if (keywordScore <= 5 && input.targetKeywords.length) fixes.push("Cover target keywords.");

  const drill = speech < delivery && speech < content
    ? "60-second pace drill"
    : delivery < content
      ? "Eye contact mirror drill"
      : "Story arc outline drill";

  return {
    total,
    speech,
    delivery,
    content,
    details: {
      wpm: wpmScore,
      fillers: fillerScore,
      pauses: pauseScore,
      eyeContact: eyeScore,
      smile: smileScore,
      headStability: headScore,
      structure: structureScore,
      repetition: repetitionScore,
      clarity: clarityScore,
      keywordCoverage: keywordScore,
    },
    topFixes: fixes.slice(0, 3),
    recommendedDrill: drill,
  };
};

File: apps/api/src/app/scoring.py
import re
from typing import List

FILLER_WORDS = ["um", "uh", "like", "you know", "actually", "basically", "literally"]
STRUCTURE_WORDS = ["first", "second", "third", "because", "therefore", "however", "so", "thus"]


def tokenize(text: str) -> List[str]:
    return [w for w in re.sub(r"[^a-z0-9\s']", " ", text.lower()).split() if w]


def count_words(segments: List[dict]) -> int:
    return sum(len(tokenize(seg["text"])) for seg in segments)


def count_fillers(segments: List[dict]) -> int:
    text = " ".join(seg["text"].lower() for seg in segments)
    total = 0
    for filler in FILLER_WORDS:
        escaped = re.escape(filler).replace(" ", r"\s+")
        pattern = re.compile(rf"\b{escaped}\b")
        total += len(pattern.findall(text))
    return total


def compute_wpm(word_count: int, duration_ms: int) -> float:
    if duration_ms <= 0:
        return 0
    return round((word_count / (duration_ms / 60000)) * 10) / 10


def score_wpm(wpm: float) -> int:
    if 120 <= wpm <= 160:
        return 18
    if 100 <= wpm < 120 or 160 < wpm <= 180:
        return 10
    return 4


def score_fillers(fillers_per_minute: float) -> int:
    if fillers_per_minute <= 2:
        return 12
    if fillers_per_minute <= 6:
        return 8
    return 3


def score_pauses(pause_events: List[int], duration_ms: int) -> int:
    if not pause_events:
        return 4
    natural = len([p for p in pause_events if 600 <= p <= 1500])
    long = len([p for p in pause_events if p > 1500])
    minutes = max(duration_ms / 60000, 0.1)
    rate = natural / minutes
    if long > max(2, minutes):
        return 4
    if 2 <= rate <= 6:
        return 10
    if rate >= 1:
        return 7
    return 5


def score_eye_contact(eye_contact_pct: float) -> int:
    if eye_contact_pct > 0.7:
        return 15
    if eye_contact_pct >= 0.4:
        return 10
    return 5


def score_smile(smile_avg: float) -> int:
    return round(7 * max(0, min(1, smile_avg)))


def score_head_stability(yaw_var: float, pitch_var: float, roll_var: float) -> int:
    total_var = yaw_var + pitch_var + roll_var
    if total_var < 5:
        return 8
    if total_var < 12:
        return 6
    return 4


def score_structure(segments: List[dict]) -> int:
    words = tokenize(" ".join(seg["text"] for seg in segments))
    count = sum(1 for w in words if w in STRUCTURE_WORDS)
    if count >= 4:
        return 10
    if count >= 2:
        return 7
    return 4


def score_repetition(segments: List[dict]) -> int:
    words = tokenize(" ".join(seg["text"] for seg in segments))
    freq = {}
    for word in words:
        freq[word] = freq.get(word, 0) + 1
    repeated = len([c for c in freq.values() if c >= 6])
    if repeated >= 3:
        return -8
    if repeated >= 1:
        return -4
    return 0


def score_clarity(segments: List[dict]) -> int:
    sentences = [s.strip() for s in re.split(r"[.!?]+", " ".join(seg["text"] for seg in segments)) if s.strip()]
    if not sentences:
        return 0
    lengths = [len(tokenize(s)) for s in sentences]
    avg = sum(lengths) / len(lengths)
    if 8 <= avg <= 20:
        return 12
    if 5 <= avg <= 25:
        return 8
    return 4


def score_keyword_coverage(segments: List[dict], keywords: List[str]) -> int:
    if not keywords:
        return 0
    text = " ".join(seg["text"].lower() for seg in segments)
    covered = sum(1 for k in keywords if k.lower() in text)
    ratio = covered / len(keywords)
    if ratio >= 0.8:
        return 8
    if ratio >= 0.5:
        return 5
    return 2


def avg(values: List[float]) -> float:
    return sum(values) / len(values) if values else 0


def variance(values: List[float]) -> float:
    if not values:
        return 0
    mean = avg(values)
    return avg([(v - mean) ** 2 for v in values])


def compute_score(segments: List[dict], metrics: List[dict], duration_ms: int, keywords: List[str]):
    word_count = count_words(segments)
    wpm = compute_wpm(word_count, duration_ms)
    filler_count = count_fillers(segments)
    minutes = max(duration_ms / 60000, 0.1)
    fillers_per_minute = filler_count / minutes
    pause_events = [m.get("pause_ms", 0) for m in metrics if m.get("pause_ms", 0) > 0]

    wpm_score = score_wpm(wpm)
    filler_score = score_fillers(fillers_per_minute)
    pause_score = score_pauses(pause_events, duration_ms)
    speech = wpm_score + filler_score + pause_score

    eye_contact_pct = avg([m.get("eye_contact", 0) for m in metrics])
    smile_avg = avg([m.get("smile", 0) for m in metrics])
    yaw_var = variance([m.get("yaw", 0) for m in metrics])
    pitch_var = variance([m.get("pitch", 0) for m in metrics])
    roll_var = variance([m.get("roll", 0) for m in metrics])

    eye_score = score_eye_contact(eye_contact_pct)
    smile_score = score_smile(smile_avg)
    head_score = score_head_stability(yaw_var, pitch_var, roll_var)
    delivery = eye_score + smile_score + head_score

    structure_score = score_structure(segments)
    repetition_score = score_repetition(segments)
    clarity_score = score_clarity(segments)
    keyword_score = score_keyword_coverage(segments, keywords)
    content = max(0, structure_score + repetition_score + clarity_score + keyword_score)

    total = max(0, min(100, speech + delivery + content))

    fixes = []
    if wpm_score <= 10:
        fixes.append("Adjust pace to 120-160 WPM.")
    if filler_score <= 8:
        fixes.append("Reduce filler words with pauses.")
    if eye_score <= 10:
        fixes.append("Increase eye contact with the lens.")
    if clarity_score <= 8:
        fixes.append("Shorten sentences for clarity.")
    if keyword_score <= 5 and keywords:
        fixes.append("Cover target keywords.")

    drill = (
        "60-second pace drill"
        if speech < delivery and speech < content
        else "Eye contact mirror drill"
        if delivery < content
        else "Story arc outline drill"
    )

    return {
        "total": total,
        "speech": speech,
        "delivery": delivery,
        "content": content,
        "details": {
            "wpm": wpm_score,
            "fillers": filler_score,
            "pauses": pause_score,
            "eyeContact": eye_score,
            "smile": smile_score,
            "headStability": head_score,
            "structure": structure_score,
            "repetition": repetition_score,
            "clarity": clarity_score,
            "keywordCoverage": keyword_score,
        },
        "topFixes": fixes[:3],
        "recommendedDrill": drill,
    }
